Add support for Tuya MMW sensors: `_TZE204_qasjif9e`, `_TZE204_sxm7l9xa`, `_TZE204_e5m9c5hl`

pgroene opened the pull request:
## Proposed change
  Add support for Tuya MMW sensor _TZE204_qasjif9e and _TZE204_sxm7l9xa



## Additional information
Added support based on example code in https://github.com/zigpy/zha-device-handlers/issues/2378#issuecomment-1666843526
#2510 

## Checklist


- [x] The changes are tested and work correctly
- [x] `pre-commit` checks pass / the code has been formatted using Black
- [ ] Tests have been added to verify that the new code works

pgroene committed:
Add support for Tuya MMW sensor _TZE204_qasjif9e

pgroene committed:
Fix build errors

Address flagged issue in build pipeline

pgroene committed:
Address formatting flagged in build pipeline

Address formatting flagged in build pipeline

pgroene committed:
Fix type indent

pgroene committed:
Use DPToAttributeMapping from mcu 

for dp_convert

pgroene committed:
fix formatting

fix formatting

pgroene commented:
How to add unittests and what kind of unittests? Or how to grt this merged?

MichalLytek commented:
> Added support based on example code in https://github.com/zigpy/zha-device-handlers/issues/2378#issuecomment-1666843526

This example code does not support `target_distance` nor `illuminance_lux`, available in Z2MQTT.
Is it possible to add this?

evgenln commented:
it supports but there are limitations. 
To get illuminance_lux, restart the HA host (not only reload ZHA).
To get target_distance it should be exposed in ZHA (is not part of this PR)

pgroene commented:
> it supports but there are limitations. To get illuminance_lux, restart the HA host (not only reload ZHA). To get target_distance it should be exposed in ZHA (is not part of this PR)

Yes, the lux is working. I tried to get the target distance working. But did not see what was missing. @evgenln if you have any pointers, i can have a look in an other pr.
Im running this code for about 2 weeks with any issues. And ordered 18 more sensors, it would be nice to run them without custom deployment actions.

pgroene committed:
Merge branch 'dev' into dev

pgroene committed:
Update ts0601_motion.py

Add support for TZE204_sxm7l9xa

pgroene committed:
Update ts0601_motion.py

fix tailing whitespaces

pgroene committed:
Update ts0601_motion.py

add missing enter

pgroene committed:
Update ts0601_motion.py

fix showing lux for _TZE204_sxm7l9xa

pgroene commented:
Merged the changes from this pr @arbrandes:
https://github.com/zigpy/zha-device-handlers/pull/2485

into this pr, it now does _TZE204_sxm7l9xa
Tested both the new devices

pgroene committed:
Update ts0601_motion.py

fix the white spaces

pgroene commented:
It seems I broke the _TZE204_qasjif9e, I'll update when I have a fix

pgroene committed:
Update ts0601_motion.py

fixed _TZE204_qasjif9e
but lost lux on _TZE204_sxm7l9xa

pgroene commented:
It seems latest version both sensors work correctly for presence and lux

ashleycawley commented:
I recently got a _TZE204_qasjif9e mmWave motion sensor and was trying to set it up. Can I assist in testing at all? I'm an amateur and coding so may not be able to help there, but willing to test and report back if I can help. The device was introduced to the ZHA network as normal and appears, just no useful entities or functions.

pgroene commented:
> I recently got a _TZE204_qasjif9e mmWave motion sensor and was trying to set it up. Can I assist in testing at all? I'm an amateur and coding so may not be able to help there, but willing to test and report back if I can help. The device was introduced to the ZHA network as normal and appears, just no useful entities or functions.

You can test the file from the PR:

Create a custom quirk dir in HA, e.g., /config/custom_zha_quirks
In configuration.yaml, point to this directory:
zha:
  custom_quirks_path: /config/custom_zha_quirks/
in this directory, create the python file with content of the quirk

edwiness10 commented:
Hello, I recently purchased two mmWave modules (_TZE204_qasjif9e and _TZE204_sxm7l9xa). With the latest change of ts0601_motion.py in the path /config/custom_zha_quirks/ I managed to configure correctly with ZHA. I can already see the entities. If there is any test I would be happy to do.

ashleycawley commented:
> You can test the file from the PR:
> 
> Create a custom quirk dir in HA, e.g., /config/custom_zha_quirks In configuration.yaml, point to this directory: zha: custom_quirks_path: /config/custom_zha_quirks/ in this directory, create the python file with content of the quirk

This is working great for me! Thank you for all your efforts with this.

pgroene commented:
@TheJulianJES what are the steps to get this merged?

typhoon2099 commented:
I've just tested this on my Home Assistant installation and it seems to be showing the controls properly for `_TZE204_qasjif9e`, but it has removed the controls for `_TZE204_ztc6ggyl`. I'm currently in the middle of rebooting to see if the controls return (assuming that the ZHA restart isn't enough.

EDIT: Restarted Home Assistant and nothing has changed (I do have luminance sensors on the `_TZE204_qasjif9e` motion detectors though):
![Screenshot_20230923_185532](https://github.com/zigpy/zha-device-handlers/assets/828104/e596808f-b7e7-461f-88e2-65f5bb80073e)
![Screenshot_20230923_185434](https://github.com/zigpy/zha-device-handlers/assets/828104/2d701cea-d177-4e83-b1f2-4ceed87b704c)

pgroene commented:
@typhoon2099 Thanks for testing, I'll check if I can find what breaks _TZE204_ztc6ggyl

TheJulianJES commented:
> what are the steps to get this merged?

Ideally, there should be some basic [unit tests](https://github.com/zigpy/zha-device-handlers/blob/dev/tests/test_tuya.py) for the added code. But I'll check if we can make an exception for this. I don't have the time to add tests myself at the moment. (EDIT: meh, it's mostly setting constraints for `AnalogOutput` clusters)

The "fake" `AnalogOutput` clusters are a hacky way of creating config entities in ZHA, but we need some fake attribute for the Tuya datapoints. The more proper way would be to put all "fake Tuya attributes" on one "fake cluster" and then create matching entities in [zha/number.py](https://github.com/home-assistant/core/blob/dev/homeassistant/components/zha/number.py) (scroll to the bottom for examples).
If we're already doing this `AnalogOutput` cluster stuff for other Tuya devices, then I guess it might be acceptable for now..

TheJulianJES commented on zhaquirks/tuya/ts0601_motion.py:
```diff
@@ -66,6 +82,94 @@ class TuyaRelativeHumidity(RelativeHumidity, TuyaLocalCluster):
+    def __init__(self, *args, **kwargs):
+        """Init."""
+        super().__init__(*args, **kwargs)
+        self._update_attribute(self.attributes_by_name["description"].id, "sensitivity")
+        self._update_attribute(self.attributes_by_name["min_present_value"].id, 1)
+        self._update_attribute(self.attributes_by_name["max_present_value"].id, 9)
+        self._update_attribute(self.attributes_by_name["resolution"].id, 1)
```
With recent-ish zigpy updates, we should be able to use `AnalogOutput.AttributeDefs.description.id` here (`self.AttributeDefs.description.id` should also work, ~~although I think I prefer the more "explicit" version~~ Use what you like better 😄).

I think it's nicer to use `AttributeDefs` everywhere.

TheJulianJES replied:
Also, can't we just use `_CONSTANT_ATTRIBUTES` in a cluster for all of this?
(Then we'd avoid updating attributes in the constructor / ``__init__ `` could be removed)

Example:
https://github.com/zigpy/zha-device-handlers/blob/7385465bd6f04f295893fb20d053433011183935/zhaquirks/terncy/cl001.py#L27-L34

TheJulianJES commented on zhaquirks/tuya/ts0601_motion.py:
```diff
@@ -115,99 +219,141 @@ class NeoMotionManufCluster(TuyaNewManufCluster):
+            0xEF03: ("min_range", t.uint32_t, True),
+            0xEF04: ("max_range", t.uint32_t, True),
+            0xEF06: ("self_test", TuyaMmwRadarSelfTest, True),
+            0xEF09: ("target_distance", t.uint32_t, True),
```
Just looking at the fake `target_distance` attribute for example, is this even used anymore?
-> see next/below comment

TheJulianJES commented on zhaquirks/tuya/ts0601_motion.py:
```diff
@@ -115,99 +219,141 @@ class NeoMotionManufCluster(TuyaNewManufCluster):
+            "self_test",
         ),
         9: DPToAttributeMapping(
-            TuyaAnalogInput.ep_attribute,
+            TuyaMmwRadarTargetDistance.ep_attribute,
             "present_value",
             lambda x: x / 100,
         ),
```
(continuing from previous/above comment:)
Here, the `present_value` on the "fake" `AnalogInput` cluster is updated with the target distance?

Is there a difference between this and the `target_distance` attribute in `TuyaMmwRadarClusterBase`?

pgroene replied:
I did update the pr for all other comments. However on the target_distance I'm not sure. I do not get it in the UI, and this comments does tell me that it is not exposed by zha:
https://github.com/zigpy/zha-device-handlers/pull/2525#issuecomment-1689649785

I did not test leaving it out (if not adding it breaks anything). 

Where to do the updates to expose the target_distance to the UI?

TheJulianJES replied:
See my [other comment](https://github.com/zigpy/zha-device-handlers/pull/2525/files#r1368052764), but the `AnalogInput` sensor only gets created in ZHA for one manufacturer from what I can see: [sensor.py#L188-L198](https://github.com/home-assistant/core/blob/4ee9a6f1301d34c2f287038ebf0c3acdc5f93517/homeassistant/components/zha/sensor.py#L188-L198)

The correct way to create a sensor for this would be by having the Tuya datapoint map to a (fake) ZCL attribute and then exposing that through ZHA with something like this: [sensor.py#L1017-L1027](https://github.com/home-assistant/core/blob/4ee9a6f1301d34c2f287038ebf0c3acdc5f93517/homeassistant/components/zha/sensor.py#L1017-L1027)

TheJulianJES reviewed (commented):
I've left some comments. IMO, the most important one would be: https://github.com/zigpy/zha-device-handlers/pull/2525/files#r1340851453

`AnalogOutput` / `AnalogInput` clusters are still a hacky way of doing this, but maybe it's fine for now. I think other Tuya stuff also does this(?)

---

And it should be verified (as best as possible) that no existing sensors are broken with the changes.

drmulligan commented:
> @typhoon2099 Thanks for testing, I'll check if I can find what breaks _TZE204_ztc6ggyl

The problem is TuyaMmwRadarOccupancyVariant2 is missing replacement endpoints 2 through 5.  Once I add the missing code, configuration seems to work for _TZE204_ztc6ggyl.  Thank you @pgroene !

zebpalmer commented:
Working for me with a `_TZE204_qasjif9e` thanks!

pgroene commented:
@TheJulianJES thanks for reviewing, I'll try to do your comments and fix the TuyaMmwRadarOccupancyVariant2  ASAP.

pgroene committed:
Fix for _TZE204_ztc6ggyl add the correct endpoints 2 to 5

review comments from @drmulligan

pgroene committed:
Fix the TuyaMmwRadarOccupancyVariant1 definitions

pgroene committed:
Explicit use for fields instead of attributes

user self.AttributeDefs.description.id instead of self.attributes_by_name["description"].id

pgroene committed:
remove references to  __init__

pgroene commented:
@drmulligan can you verify that I fix the _TZE204_ztc6ggyl sensor with the latest version? I do not have that sensor

lukkarm1 commented:
I have an another variant, `_TZE204_e5m9c5hl` (Sold as WZ-M100-W), but after reading related discussions, it seems to be similar to `_TZE204_sxm7l9xa`.

And, indeed, after testing the code by adding the model in variant 3, I get the device working with occupancy and illuminance

drmulligan commented:
> @drmulligan can you verify that I fix the _TZE204_ztc6ggyl sensor with the latest version? I do not have that sensor

Still works, thank you.

pgroene commented:
@lukkarm1 can you control the settings correctly for the _TZE204_e5m9c5hl? If, so can you confirm that before I add the sensor to the list?

pgroene committed:
Merge branch 'zigpy:dev' into dev

pgroene committed:
remove unintended whitespace

lukkarm1 commented:
> @lukkarm1 can you control the settings correctly for the _TZE204_e5m9c5hl? If, so can you confirm that before I add the sensor to the list?

Sure, I realized now that I forgot to inlude that information. I'm not sure how the not working settings would present themselves, but changing the values in UI seems to work and I checked that they match the values in `manage zigbee device` relevant cluster's `present_value`s

pgroene committed:
Add _TZE204_e5m9c5hl

_TZE204_e5m9c5hl should be the same as variant3

pgroene commented:
@TheJulianJES I hope I updated the code to be in an acceptable state for merging. I also added an extra device _TZE204_e5m9c5hl

jcesclapez commented:
Hello, 
I have the _TZE204_qasjif9e but the Occupancy sensor is always showing "Not detected", Illuminance sensor is ok and the bars too.

![imagen](https://github.com/zigpy/zha-device-handlers/assets/41897908/658cfb4d-d8dd-4e74-b2e8-d26adb2ba6d2)


What can i do? I have de celling version not the mount wall.

evgenln commented:
Answer clear from the settings. What you expect with lowest sensitivity, same and min and max range

jcesclapez commented:
> Answer clear from the settings. What you expect with lowest sensitivity, same and min and max range

Thanks for answer. 


Those are the options that remained after trying,

![imagen](https://github.com/zigpy/zha-device-handlers/assets/41897908/de2d19b1-0fd8-4c62-b1f8-0e3c2e30efa7)

With these ranges it remains undetected and this right now under the sensor..

evgenln commented:
From my personal experience ranges for each model are bit different, e.g. min range cannot be 0 or max cannot be 12 m. In some devices it's easy to check writing values to appropriate property in mange device and then read back checking if value was applied, in some value resets to last  Ok with some time. Try set min range about 1m and max to about 5m (if you test on table). Sensitive 7-8. This quirk is too universal so count devices as same based on datapoints match but does not handle acceptable values for every manufacturer

jcesclapez commented:
> From my personal experience ranges for each model are bit different, e.g. min range cannot be 0 or max cannot be 12 m. In some devices it's easy to check writing values to appropriate property in mange device and then read back checking if value was applied, in some value resets to last Ok with some time. Try set min range about 1m and max to about 5m (if you test on table). Sensitive 7-8. This quirk is too universal so count devices as same based on datapoints match but does not handle acceptable values for every manufacturer

Thanks for answering so quickly. 

I have tried the settings you have suggested but the state still does not change, in addition to trying other combinations as you have mentioned.

![imagen](https://github.com/zigpy/zha-device-handlers/assets/41897908/e13b4323-9905-4570-952b-898d0059fe1c)

I'm starting to think that the sensor may have been broken... Is there a way to debug this and see what the sensor picks up?

evgenln commented:
Not sure, you can try review logs for errors/events but it requires related knowledge. In your case probably most easy way will be to try alternative handlers like Z2m

MattWestb commented:
Add:
```
logger:
  default: info
  logs:
    homeassistant.core: info
    homeassistant.components.zha: debug
    zigpy: debug
    zhaquirks: debug
    custom_zha_quirks: debug
```
in you  HA config and restart  HA and you shall getting the data the quirks is getting / sending.

MasterDevwi commented:
I recently purchased a _TZE204_e5m9c5hl and _TZE204_ijxvkhd0 and I've been trying to set them up.

I was able to get the entities for the _TZE204_e5m9c5hl to show up by copying the contents of zhaquirks/tuya/ts0601_motion.py into a new file called TS0601_TZE204_e5m9c5hl.py in my custom_zha_quirks directory. 

<img width="499" alt="image" src="https://github.com/zigpy/zha-device-handlers/assets/5544757/c35ba673-f2c2-48d4-84a1-bb8e7f49880b">

Unfortunately, the occupancy sensor always says "Clear." I cannot for the life of me get it to detect anything, even after changing the values. Any idea how to get it working?

I've had even less luck with the _TZE204_ijxvkhd0.

TheJulianJES commented on zhaquirks/tuya/ts0601_motion.py:
```diff
@@ -66,6 +82,74 @@ class TuyaRelativeHumidity(RelativeHumidity, TuyaLocalCluster):
+    }
+
+
+class TuyaMmwRadarTargetDistance(TuyaAttributesCluster, AnalogInput):
+    """AnalogInput cluster for target distance."""
+
+    _CONSTANT_ATTRIBUTES = {
+        AnalogOutput.AttributeDefs.description.id: "target_distance",
+        AnalogOutput.AttributeDefs.engineering_units.id: 118,  # 31: meters
+    }
```
I previously thought `target_distance` was supposed to be set from HA as well, but since this is an `AnalogInput` cluster, it means that `target_distance` comes from the sensor and should be exposed (as an output) to ZHA, right?
That's not possible without code changes to ZHA itself yet.

---

I'm currently thinking if we should already move all `AnalogOutput` and `AnalogInput` stuff directly to ZHA (to expose it properly in HA). Otherwise, it probably needs to be done in the future.

For how to set up configuration entities properly in ZHA:
- https://github.com/home-assistant/core/blob/4ee9a6f1301d34c2f287038ebf0c3acdc5f93517/homeassistant/components/zha/number.py#L562-L579
- This would require https://github.com/home-assistant/core/pull/102482 though (to not duplicate all manufactuer/models into ZHA again)

It's not ideal that custom entities can only be created using ZHA at the moment, and it should (hopefully) be possible with quirks directly at some point in the future. Using `AnalogOutput` is a hack though. But it's already done for some other Tuya quirks and I'd really like to get these sensors supported.

I'll need to see if the referenced PR gets merged before the next HA release (sadly doubt that). I'll get back to you/this PR. Sorry for all the waiting. Thanks for all the work on this PR so far!

TheJulianJES replied:
The ideal way forward for this PR would be:
- to remove the `AnalogOutput` clusters (which are used as an input to the device)
- to remove the `AnalogInput` cluster (which is used as an output from the device to HA)
- map all Tuya datapoints to custom attributes on one cluster (instead of on individual `AnalogOutput` clusters)
- assign a `quirk_id` to the quirks like here: [#2674](https://github.com/zigpy/zha-device-handlers/pull/2674/files)
- add custom entities in HA similar to this: [zha/number.py#L562-L579](https://github.com/home-assistant/core/blob/a59076d140e316af686a9ed705281e0d6255ac9c/homeassistant/components/zha/number.py#L562-L579) but using `quirk_id` like here: [zha/select.py#L252-L261](https://github.com/home-assistant/core/blob/a59076d140e316af686a9ed705281e0d6255ac9c/homeassistant/components/zha/select.py#L252-L261)
- initialize the custom attributes (so they're read once upon pairing, else the entity doesn't get created) like here: [zha/core/cluster_handlers/manufacturerspecific.py#L64-L80](https://github.com/home-assistant/core/blob/a59076d140e316af686a9ed705281e0d6255ac9c/homeassistant/components/zha/core/cluster_handlers/manufacturerspecific.py#L64-L80)

That way, the entities are added properly in ZHA. Eventually, I hope that we'll get the functionality to create entities directly via quirks, but this isn't possible at the moment yet.

TheJulianJES replied:
> That would negate the need to write quirks?

No, the Tuya datapoints still have to be mapped to (fake) attributes on a (fake) cluster, as that's the only way we can get that information to ZHA.
However, entity matches in ZHA also need to be added for those attributes.

pgroene replied:
> The ideal way forward for this PR would be:
> 
> * to remove the `AnalogOutput` clusters (which are used as an input to the device)
> * to remove the `AnalogInput` cluster (which is used as an output from the device to HA)
> * map all Tuya datapoints to custom attributes on one cluster (instead of on individual `AnalogOutput` clusters)
> * assign a `quirk_id` to the quirks like here: [#2674](https://github.com/zigpy/zha-device-handlers/pull/2674/files)
> * add custom entities in HA similar to this: [zha/number.py#L562-L579](https://github.com/home-assistant/core/blob/a59076d140e316af686a9ed705281e0d6255ac9c/homeassistant/components/zha/number.py#L562-L579) but using `quirk_id` like here: [zha/select.py#L252-L261](https://github.com/home-assistant/core/blob/a59076d140e316af686a9ed705281e0d6255ac9c/homeassistant/components/zha/select.py#L252-L261)
> * initialize the custom attributes (so they're read once upon pairing, else the entity doesn't get created) like here: [zha/core/cluster_handlers/manufacturerspecific.py#L64-L80](https://github.com/home-assistant/core/blob/a59076d140e316af686a9ed705281e0d6255ac9c/homeassistant/components/zha/core/cluster_handlers/manufacturerspecific.py#L64-L80)
> 
> That way, the entities are added properly in ZHA. Eventually, I hope that we'll get the functionality to create entities directly via quirks, but this isn't possible at the moment yet.

I'll have a look at these items the coming weeks

- [ ] to remove the `AnalogOutput` clusters (which are used as an input to the device)
- [ ] to remove the `AnalogInput` cluster (which is used as an output from the device to HA)
- [ ]  map all Tuya datapoints to custom attributes on one cluster (instead of on individual `AnalogOutput` clusters)
- [X] assign a `quirk_id` to the quirks like here: [#2674](https://github.com/zigpy/zha-device-handlers/pull/2674/files)
- [ ]  add custom entities in HA similar to this: [zha/number.py#L562-L579](https://github.com/home-assistant/core/blob/a59076d140e316af686a9ed705281e0d6255ac9c/homeassistant/components/zha/number.py#L562-L579) but using `quirk_id` like here: [zha/select.py#L252-L261](https://github.com/home-assistant/core/blob/a59076d140e316af686a9ed705281e0d6255ac9c/homeassistant/components/zha/select.py#L252-L261)
- [ ] initialize the custom attributes (so they're read once upon pairing, else the entity doesn't get created) like here: [zha/core/cluster_handlers/manufacturerspecific.py#L64-L80](https://github.com/home-assistant/core/blob/a59076d140e316af686a9ed705281e0d6255ac9c/homeassistant/components/zha/core/cluster_handlers/manufacturerspecific.py#L64-L80)

pgroene replied:
@TheJulianJES  I did  assign a `quirk_id`, if I copy the file quirk_ids.py in my custom quirks directory, I do not get the import in the  ts0601_motion.py. How do I test this?

pgroene replied:
@TheJulianJES I tried to add the quirk_id, however how to test it?

Would it not be a better plan to merge in the working version of the quirk and than work to improve it with smaller commits. It seems doing everything at once will take a long time, and doing changes over multiple repositories makes it a complex task to do at once where I do not have the time to do test and set it up all at once. This would also allow other people to contribute and add more related devices

I would suggest to rollback to the working version and that do separate commits to get the other changes in.

racedo replied:
+1

TheJulianJES replied:
> if I copy the file quirk_ids.py in my custom quirks directory, I do not get the import in the ts0601_motion.py. How do I test this?

You'd need to change the import to something like `from . import xxx`, but the `quirk_id` is just a string, so you don't need to import it from the quirks id file, and can just set it like `quirk_id = "tuya.mmw_radar"`.
In the long run, it's easier to test this with completely replacing the `zhaquirks` package with "a custom one" instead of using custom quirks.
But again, you can just not import it and set `quirk_id = "tuya.mmw_radar"` everywhere in your custom quirk file (or define `TUYA_MMWRADAR = "tuya.mmw_radar"` in your custom quirk file for now).
This can be cleaned up just before merging.

> I tried to add the quirk_id, however how to test it?

It won't do anything on its own, it's just to make it easy to match the quirk to custom entities In ZHA.

> Would it not be a better plan to merge in the working version of the quirk and than work to improve it with smaller commits.

So, the issue with the current version is that we use "hacky" `AnalogInput` and `AnalogOutput` clusters to expose entities in ZHA. If we later decide to properly implement custom entities in ZHA and remove the (fake) `AnalogInput` and `AnalogOutput` clusters from this quirk, the new entities will differ from the old ones.
Entity IDs will change, the old entities will be "no longer provided" but stay in HA until deleted.

Like, I'd be fine to merge the quirk without any configuration entities for now. But I don't think we should include `AnalogInput` / `AnalogOutput` stuff, just to remove it again later (which would be somewhat of a breaking change).

 > It seems doing everything at once will take a long time, and doing changes over multiple repositories makes it a complex task to do at once where I do not have the time to do test and set it up all at once. This would also allow other people to contribute and add more related devices

Of course I understand this though.

> I would suggest to rollback to the working version

Hmm, what changes broke something? Like, what isn't working correctly? Just the quirk ID import?

pgroene commented:
> Hello, I have the _TZE204_qasjif9e but the Occupancy sensor is always showing "Not detected", Illuminance sensor is ok and the bars too.
> 
> ![imagen](https://user-images.githubusercontent.com/41897908/277158481-658cfb4d-d8dd-4e74-b2e8-d26adb2ba6d2.png)
> 
> What can i do? I have de celling version not the mount wall.

@jcesclapez Are you sure you are using the correct files? I would expect TuyaMmwRadarOccupancyVariant2 to show up at the quirk and not the name you are seeing right now in the screenshot.

jcesclapez commented:
> > Hello, I have the _TZE204_qasjif9e but the Occupancy sensor is always showing "Not detected", Illuminance sensor is ok and the bars too.
> > ![imagen](https://user-images.githubusercontent.com/41897908/277158481-658cfb4d-d8dd-4e74-b2e8-d26adb2ba6d2.png)
> > What can i do? I have de celling version not the mount wall.
> 
> @jcesclapez Are you sure you are using the correct files? I would expect TuyaMmwRadarOccupancyVariant2 to show up at the quirk and not the name you are seeing right now in the screenshot.

Hello @pgroene, 

In my configuration file I have these lines

zha:
  enable_quirks: true
  custom_quirks_path: /config/custom_zha_quirks/
  database_path: /config/zigbee.db

And the file I have in this path /config/custom_zha_quirks/ is ts0601_montion.py

(https://raw.githubusercontent.com/zigpy/zha-device-handlers/c7b130a20614e3e364c9fea023d632014a5726d1/zhaquirks/tuya/ts0601_motion.py)

If I remove this py file, then it only show the entities LQI and RSSI.


Is correct?


Regards

EDIT: After delete the folder __pycache__ and the device from zha and reboot it started working, sorry and thanks for your work, very appreciate.

cam-el commented:
I have been following this for awhile. I have a "TS0601 by _TZE204_qasjif9e " device and have the ts0601_motion.py file loading from [zha-device-handlers](https://raw.githubusercontent.com/zigpy/zha-device-handlers/c7b130a20614e3e364c9fea023d632014a5726d1/zhaquirks/tuya/ts0601_motion.py) with a cleared "__pycache__" folder and multiple reboots
![Screenshot from 2023-10-29 13-38-54](https://github.com/zigpy/zha-device-handlers/assets/77422102/189300ef-246e-47e7-9e3d-a1b19d02598c)

Moving the sliders does not do much. The device is close to a working zigbee router.
No illuminance sensor has shown up and the occupancy does not change state. not sure how to check that independently. 

The model is like a hockey pock, not the ceiling mounted version

cam-el commented:
Thanks @3dbluebot . I am keen to try a few more things and will give that a go. I saw you and others trying that but tried to stay to the same model I had. Will try a few more permutations.

pgroene commented:
@3dbluebot I want to spent the time first on getting the current version in, and than we can add a new definition. Mostly because I can not test the changes my self. I saw the mapping the mqtt handler, so it is probably not so hard to add (only change a few numbers for a new variant).

typhoon2099 commented:
I've just had a go with my Home Assistant instance and both `_TZE204_qasjif9e` and `_TZE204_ztc6ggyl` are working correctly again!

TheJulianJES commented:
To make sure my new comment is seen (discussion further up), I'm linking it here too: https://github.com/zigpy/zha-device-handlers/pull/2525#discussion_r1402934740

pgroene committed:
Merge branch 'zigpy:dev' into dev

pgroene committed:
Update quirk_ids.py

Add id for tuya MMW radar devices

pgroene committed:
Update ts0601_motion.py

pgroene committed:
Update quirk_ids.py

pgroene committed:
Update ts0601_motion.py

Coysh commented:
Would be interested to know if I can use this as a custom quirk? I'm relatively new to quirks, but have a _TZE204_qasjif9e and want to get that working in my ZHA Home Assistant.

Or will this be merged into the main Home Assistant ZHA codebase at some point?

SavoryLikeBacon commented:
I just picked up a few and tried the current code with qasjif9e and its doesn't work for me. I have a sxm7l9xa already working and the current made it so the controls were not changeable so I went back.

I also really appreciate the work being done for this sensors. the one I have is SOOO cool that I got 3 more,...but they are different models...

Coysh commented:
@SavoryLikeBacon - So you've tried this with qasjif9e and it's not working?

pgroene commented:
@Coysh @SavoryLikeBacon Can you use this older version of the quirk:
https://[raw.githubusercontent.com/zigpy/zha-device-handlers/c7b130a20614e3e364c9fea023d632014a5726d1/zhaquirks/tuya/ts0601_motion.py](https://raw.githubusercontent.com/zigpy/zha-device-handlers/c7b130a20614e3e364c9fea023d632014a5726d1/zhaquirks/tuya/ts0601_motion.py)

This version is working, the current version is not working because we try to make some breaking changes that have to go on myltiple places.

SavoryLikeBacon commented:
Hi @Coysh and @pgroene,

I appreciate the work. 
I did try the code above and my qasjif9e _**does _** work now! Thanks!

>

observethenoyes commented:
> @Coysh @SavoryLikeBacon Can you use this older version of the quirk: https://[raw.githubusercontent.com/zigpy/zha-device-handlers/c7b130a20614e3e364c9fea023d632014a5726d1/zhaquirks/tuya/ts0601_motion.py](https://raw.githubusercontent.com/zigpy/zha-device-handlers/c7b130a20614e3e364c9fea023d632014a5726d1/zhaquirks/tuya/ts0601_motion.py)
> 
> This version is working, the current version is not working because we try to make some breaking changes that have to go on myltiple places.

Confirmed working. I have a _TZE204_e5m9c5hl and can see the controls and occupency sensor in HA.

Coysh commented:
> @Coysh @SavoryLikeBacon Can you use this older version of the quirk: https://[raw.githubusercontent.com/zigpy/zha-device-handlers/c7b130a20614e3e364c9fea023d632014a5726d1/zhaquirks/tuya/ts0601_motion.py](https://raw.githubusercontent.com/zigpy/zha-device-handlers/c7b130a20614e3e364c9fea023d632014a5726d1/zhaquirks/tuya/ts0601_motion.py)
> 
> This version is working, the current version is not working because we try to make some breaking changes that have to go on myltiple places.

Yep - this one is working for me, see attached:
<img width="1023" alt="CleanShot 2023-12-22 at 16 39 51@2x" src="https://github.com/zigpy/zha-device-handlers/assets/5764356/fb70a56e-4151-4e32-8081-0d8a1dae91e8">


Not sure if it should be logging the history of the controls though, but I'm not experienced in these things.

Goaheadz commented:
So I just received a new mmwave sensor TS0601 by _TZE204_muvkrjr5 from Aliexpress.
I followed the above steps copying the TS0601.py file, but only LDI and RSSI is showed as expected (restart also done) cause no one talks about this sensor variant.

I am new to all this. I thougth integration would be straight, but I guess I need a quirk file to make parameters available in Home Assistant :(

Can someone help?

![image](https://github.com/zigpy/zha-device-handlers/assets/13796554/8f6e93c0-fae9-4dbe-9b92-a2b3dcf22ce4)

https://pt.aliexpress.com/item/1005006412041826.html?spm=a2g0o.order_list.order_list_main.29.52afcaa4H5VTTL&gatewayAdapt=glo2bra

Goaheadz commented:
> TZE204_muvkrjr5

INFO (MainThread) [zigpy.endpoint] [0xb905:1] Discovered endpoint information: SizePrefixedSimpleDescriptor(endpoint=1, profile=260, device_type=81, device_version=1, input_clusters=[4, 5, 61184, 0], output_clusters=[25, 10])

thefunkygibbon commented:
> Hi @Coysh and @pgroene,
> 
> I appreciate the work. I did try the code above and my qasjif9e _**does _** work now! Thanks!
> 
> >

does yours report motion/presence/illuminance and detected distance?  i have TZE204_qasjif9e but I can't seem to get anything to work other than the presence and illum and thats using https://fixtse.com/blog/zy-m100-full-zha-support ,  I can't get any of them to work from ones i've found in here or other places on github :-|

thefunkygibbon commented:
just to add/clarify, I'm have used the py code above, replaced the signature to reflect my model but it never seems to pick up on this quirk, restarted ha and re-paired the sensor but still nothing.  just picks up default quirk.  I'm unsure why.  nothing is in logs saying it can't parse it.  it should at least pick it up and use it even if it doesnt work

Goaheadz commented:
> just to add/clarify, I'm have used the py code above, replaced the signature to reflect my model but it never seems to pick up on this quirk, restarted ha and re-paired the sensor but still nothing. just picks up default quirk. I'm unsure why. nothing is in logs saying it can't parse it. it should at least pick it up and use it even if it doesnt work

First you need to make sure, that quirk was loaded:
![image](https://github.com/zigpy/zha-device-handlers/assets/13796554/7494be43-8d7e-4f61-a5de-930ce24c84ce)

thefunkygibbon commented:
yes thats what i'm saying, it wasn't loaded; it is using the default quirk still

Goaheadz commented:
You have the .py file in this location and those lines of code in configuration.yaml?
![image](https://github.com/zigpy/zha-device-handlers/assets/13796554/5ffe976a-a58a-403f-91d2-047b47cff797)
![image](https://github.com/zigpy/zha-device-handlers/assets/13796554/98326961-d482-4832-9d5c-17ee0b0ef885)
If you have all this, and after a HA restart it doesn't load the costume quirk, then you must have a mistype somewhere.

Goaheadz commented:
And I think you should delete the _pycache_ folder, so HA can rebuild it with new quirk.

thefunkygibbon commented:
yes, all those aspects are fine.  i'm using other quirks for other devices.  and using a different quirk for this device works,  just this quirk doesnt and i REALLY need the target distance aspect to work,  and i didnt know i could get the temp from it too

thefunkygibbon commented:
to be clear, I changed

    signature = {
        MODELS_INFO: [("_TZE204_yensya2c", "TS0601")],


and replaced with 

    signature = {
        MODELS_INFO: [("_TZE204_qasjif9e", "TS0601")],


and copied it as is.  so I don't see how i could have typo'd

thefunkygibbon commented:
i'm sure it is something to do with the code so that it isn't just that line of the py file signature that causes HA to associate the quirk with it. 
 thanks for your help.  hopefully one of the others who have used this with my model can chime in (or @3dbluebot who wrote it)

llucax commented:
> @Coysh @SavoryLikeBacon Can you use this older version of the quirk: https://[raw.githubusercontent.com/zigpy/zha-device-handlers/c7b130a20614e3e364c9fea023d632014a5726d1/zhaquirks/tuya/ts0601_motion.py](https://raw.githubusercontent.com/zigpy/zha-device-handlers/c7b130a20614e3e364c9fea023d632014a5726d1/zhaquirks/tuya/ts0601_motion.py)
> 
> This version is working, the current version is not working because we try to make some breaking changes that have to go on myltiple places.

In case it is useful, I can confirm this version works for me too for _TZE204_e5m9c5hl. Awesome, thanks! I hope this PR is sorted out soon and the support is merged.

nirajbhatt7 commented:
> > @Coysh @SavoryLikeBacon Can you use this older version of the quirk: https://[raw.githubusercontent.com/zigpy/zha-device-handlers/c7b130a20614e3e364c9fea023d632014a5726d1/zhaquirks/tuya/ts0601_motion.py](https://raw.githubusercontent.com/zigpy/zha-device-handlers/c7b130a20614e3e364c9fea023d632014a5726d1/zhaquirks/tuya/ts0601_motion.py)
> > This version is working, the current version is not working because we try to make some breaking changes that have to go on myltiple places.
> 
> In case it is useful, I can confirm this version works for me too for _TZE204_e5m9c5hl. Awesome, thanks! I hope this PR is sorted out soon and the support is merged.

This version worked for me with following
* Only detects when someone is at 5 ft / 1.5 meter (Max is set to maximum and minimum is set to 0)
* After someone leaves the room, the "Away" triggers after about 20 to 30 seconds
* LUX sensor/values are not available

rrooggiieerr commented:
It would be great if these Tuya MMW Sensors would have native support in ZHA and no custom quirk was needed. Is there any reason this quirk is not incorporated in ZHA? What's holding it back?

thefunkygibbon commented:
> It would be great if these Tuya MMW Sensors would have native support in ZHA and no custom quirk was needed. Is there any reason this quirk is not incorporated in ZHA? What's holding it back?

mainly because Tuya don't adhere to the Zigbee standard.

TheJulianJES commented:
[Documentation](https://github.com/zigpy/zha-device-handlers/pull/3019/files) isn't quite ready yet, but all "fake" `AnalogInput` cluster for settings can be replaced with quirks v2 config entities.

pgroene commented:
> to be clear, I changed
> 
> ```
> signature = {
>     MODELS_INFO: [("_TZE204_yensya2c", "TS0601")],
> ```
> 
> and replaced with
> 
> ```
> signature = {
>     MODELS_INFO: [("_TZE204_qasjif9e", "TS0601")],
> ```
> 
> and copied it as is. so I don't see how i could have typo'd

I do not think those are the same, just replacing the signature will not get it to work. You have to do the mapping for the yensya2c to get it to work.

Cybermb commented:
What is status for this PR?
Are there plans to add fretting, presence, motion, distance detections? As in current state and in terms of functionality, it functions as a cheap PIR sensor, just more precise and a bit more control (with this quirk). There are no states for person standing, sitting, walking, jumping, etc...

pgroene commented:
I have moved to zigbee2mqtt for these devices. After there was issue in HA where these devices took down my network (ZHA) plus the time it takes to get all requirements to make this ready for merging, I have spend my time re-pairing my devices once to MQTT and the work fine.

Coysh commented:
I've finally got around to trying this and encountering a weird performance issue. Has anyone else experienced this? There is no one in or around the room, but the device keeps becoming unavailable?

Sensitivity set to 6 and distance 600

<img width="308" alt="CleanShot 2024-07-07 at 20 35 45@2x" src="https://github.com/zigpy/zha-device-handlers/assets/5764356/140a1935-d986-4fa6-81b9-3837b6aab219">

thefunkygibbon commented:
anyone got a better quirk for these devices yet?

prairiesnpr commented:
These were added as v2 quirks.

thefunkygibbon commented:
> These were added as v2 quirks.

added where and how (and when)?  are they built in to standard HA now so doesnt need a quirk... or??

prairiesnpr commented:
See https://github.com/zigpy/zha-device-handlers/blob/4f64c181d6a98c7c733bccfb6a7e52e47ad9f146/zhaquirks/tuya/ts0601_motion.py#L314

Will be part of the next HA release.
